### PR TITLE
Block Vacancy and gas price tracking

### LINKED ISSUE #4494: Track block vacancy to adjust the gas price based on a ceiling and floor value
Based on the current utilization of the block space within a given era, we should adjust the gas price for the following era and adjust the gas price to reflect the demand of block space within the chain.

### PR DESCRIPTION
**Pull request by darthsiroftardis**
CHANGELOG:

- Added tracking to the contract runtime which determines a gas price for the next era based on the utilization of the chain in the current era
- Added a filter mechanism to the `transaction_buffer` which filters out transactions and deploys that specify a tolerance less than the current gas price
- Added the current gas price of a given era to the block header

Closes #4494

### CODE REVIEW

File: node/src/components/contract_runtime/utils.rs [hunk 1]
```diff
@@ -648,3 +1004,27 @@ mod tests {
...
+        self.gas_price
+    }
+
+    pub(crate) fn maybe_gas_price_for_era_id(&self, era_id: EraId) -> Option<u8> {
+        if self.era_id == era_id {
+            return Some(self.gas_price);
+        }
+
+        None
+    }
```
Comment (EdHastingsCasperAssociation) [inline node/src/components/contract_runtime/utils.rs]: there is a types.rs file in the contract runtime folder where this type can live

File: types/src/transaction/transaction_v1/transaction_v1_header.rs [hunk 2]
```diff
@@ -151,6 +151,21 @@ impl TransactionV1Header {
...
+    pub fn gas_tolerance(&self) -> u64 {
+        match self.pricing_mode {
+            PricingMode::Classic { gas_price, .. } => gas_price,
+            PricingMode::Fixed {
+                gas_price_tolerance,
+                ..
+            } => gas_price_tolerance,
+            PricingMode::Reserved { .. } => {
+                // TODO: Change this when reserve gets implemented.
+                0u64
```
Comment (EdHastingsCasperAssociation) [inline types/src/transaction/transaction_v1/transaction_v1_header.rs:164]: i think 0 actually is the correct value...they already paid so should have no tolerance for paying more / again?

File: node/src/components/contract_runtime/utils.rs [hunk 3]
```diff
@@ -52,6 +57,282 @@ where
         .expect("task panicked")
 }
 
+// #[allow(clippy::too_many_arguments)]
+// pub(super) async fn exec_or_requeue<REv>(
```
Comment (EdHastingsCasperAssociation) [inline node/src/components/contract_runtime/utils.rs]: :dagger:

File: node/src/components/contract_runtime/operations.rs [hunk 4]
```diff
@@ -67,6 +69,12 @@ pub fn execute_finalized_block(
         });
     }
 
+    if executable_block.era_report.is_some() && next_era_gas_price.is_none() {
+        return Err(BlockExecutionError::FailedToGetNewEraGasPrice {
+            era_id: executable_block.era_id.successor(),
+        });
+    }
```
Comment (EdHastingsCasperAssociation) [inline node/src/components/contract_runtime/operations.rs:76]: this is basically fatal, isn't it?

Comment (darthsiroftardis) [reply node/src/components/contract_runtime/operations.rs:76]: Yes, but it gets translated into a fatal by the calling logic

File: node/src/components/contract_runtime/utils.rs (hunk 3 above)
Comment (darthsiroftardis) [reply node/src/components/contract_runtime/utils.rs]: Removed

File: types/src/transaction/transaction_v1/transaction_v1_header.rs (hunk 2 above)
Comment (darthsiroftardis) [reply types/src/transaction/transaction_v1/transaction_v1_header.rs:164]: Okay, good, I just put the TODO as a reminder to address this

File: node/src/components/contract_runtime/utils.rs (hunk 1 above)
Comment (darthsiroftardis) [reply node/src/components/contract_runtime/utils.rs]: Moved

**Review by EdHastingsCasperAssociation — APPROVED** [review]
optimistic approval post discussed changes

**Pull request merged**